# **Goals : **
Columns containing categories with only two factors must be stored as Booleans (bool).

Columns containing integers only must be stored as 32-bit integers (int32).

Columns containing floats must be stored as 16-bit floats (float16).

Columns containing nominal categorical data must be stored as the category data type.

Columns containing ordinal categorical data must be stored as ordered categories, and not mapped to numerical values, with an order that reflects the natural order of the column.

The DataFrame should be filtered to only contain students with 10 or more years of experience at companies with at least 1000 employees, as their recruiter base is suited to more experienced professionals at enterprise companies.

Dataset : customer_train.csv, which is a subset of a data science training provider customer dataset, so you can create a proof-of-concept of a much more efficient storage solution. The dataset contains anonymized student information, and whether they were looking for a new job or not during training.

Datasource: Datacamp

In [31]:
# Import necessary libraries
import pandas as pd

# Load the dataset
df = pd.read_csv("customer_train.csv")
# View the dataset
df.head()

student_id      city  city_development_index gender  \
0        8949  city_103                   0.920   Male   
1       29725   city_40                   0.776   Male   
2       11561   city_21                   0.624    NaN   
3       33241  city_115                   0.789    NaN   
4         666  city_162                   0.767   Male   

       relevant_experience enrolled_university education_level  \
0  Has relevant experience       no_enrollment        Graduate   
1   No relevant experience       no_enrollment        Graduate   
2   No relevant experience    Full time course        Graduate   
3   No relevant experience                 NaN        Graduate   
4  Has relevant experience       no_enrollment         Masters   

  major_discipline experience company_size    company_type last_new_job  \
0             STEM        >20          NaN             NaN            1   
1             STEM         15        50-99         Pvt Ltd           >4   
2             STEM          5          NaN             NaN        never   
3  Business Degree         <1          NaN         Pvt Ltd        never   
4             STEM        >20        50-99  Funded Startup            4   

   training_hours  job_change  
0              36           1  
1              47           0  
2              83           0  
3              52           1  
4               8           0

In [32]:
df.memory_usage().sum()

800032

In [33]:
df.shape

(7142, 14)

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7142 entries, 0 to 7141
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   student_id              7142 non-null   int64  
 1   city                    7142 non-null   object 
 2   city_development_index  7142 non-null   float64
 3   gender                  5528 non-null   object 
 4   relevant_experience     7142 non-null   object 
 5   enrolled_university     6996 non-null   object 
 6   education_level         6984 non-null   object 
 7   major_discipline        6097 non-null   object 
 8   experience              7125 non-null   object 
 9   company_size            4918 non-null   object 
 10  company_type            4851 non-null   object 
 11  last_new_job            6982 non-null   object 
 12  training_hours          7142 non-null   int64  
 13  job_change              7142 non-null   int64  
dtypes: float64(1), int64(3), object(10)
memo

In [35]:
df.describe()

student_id  city_development_index  training_hours   job_change
count   7142.000000             7142.000000     7142.000000  7142.000000
mean   16850.314758                0.828470       65.819938     0.251050
std     9661.806314                0.122943       60.126646     0.433648
min        2.000000                0.448000        1.000000     0.000000
25%     8525.500000                0.740000       23.000000     0.000000
50%    16820.500000                0.899000       48.000000     0.000000
75%    25264.250000                0.920000       89.000000     1.000000
max    33380.000000                0.949000      336.000000     1.000000

In [36]:
df.isna().sum()

student_id                   0
city                         0
city_development_index       0
gender                    1614
relevant_experience          0
enrolled_university        146
education_level            158
major_discipline          1045
experience                  17
company_size              2224
company_type              2291
last_new_job               160
training_hours               0
job_change                   0
dtype: int64

In normale cases we have to deal with the missing values but since it won't effect our work this time , we will leave them

In [37]:
df.duplicated().sum()

0

In [38]:
df['last_new_job'].value_counts()

last_new_job
1        2996
>4       1191
2        1078
never     934
3         393
4         390
Name: count, dtype: int64

In [39]:
last_cateogry=['never','1','3','4','>4']
df['last_new_job']=df['last_new_job'].astype('category')
df['last_new_job']=df['last_new_job'].cat.set_categories(new_categories=last_cateogry,ordered=True)

In [40]:
df['company_type']=df['company_type'].astype('category')

In [41]:
df['company_size'].value_counts()

company_size
50-99        1162
100-499       953
10000+        745
10-49         528
1000-4999     525
<10           482
500-999       324
5000-9999     199
Name: count, dtype: int64

In [42]:
size_categorie=['<10','10-49','50-99','100-499','500-999','1000-4999','5000-9999','10000+']
df['company_size']=df['company_size'].astype('category')
df['company_size']=df['company_size'].cat.set_categories(new_categories=size_categorie,ordered=True)

In [43]:
df['experience'].value_counts()

experience
>20    1187
5       548
3       516
4       513
6       443
2       437
7       404
9       376
10      352
8       277
11      264
15      257
14      223
1       210
<1      197
16      187
12      185
13      143
19      119
18      117
17      114
20       56
Name: count, dtype: int64

In [44]:
experience_cat=['<1','1','2','3','4','5','6','7','8','9','10','11','12','13','14','15','16','17','18','19','20','>20']
df['experience']=df['experience'].astype('category')
df['experience']=df['experience'].cat.set_categories(new_categories=experience_cat,ordered=True)

In [45]:
df['major_discipline'].value_counts()

major_discipline
STEM               5407
Humanities          247
Other               142
Business Degree     122
Arts                101
No Major             78
Name: count, dtype: int64

In [46]:
df['major_discipline']=df['major_discipline'].astype('category')

In [47]:
df['education_level'].value_counts(ascending=False)

education_level
Graduate          4338
Masters           1607
High School        766
Phd                159
Primary School     114
Name: count, dtype: int64

In [48]:
education_categories=["Primary School","High School","Graduate","Masters","Phd"]
df['education_level']=df['education_level'].astype('category')
df['education_level']=df['education_level'].cat.set_categories(new_categories=education_categories,ordered=True)

In [49]:
df['city']=df['city'].astype('category')
df['enrolled_university']=df['enrolled_university'].astype('category')

In [50]:
univ_cat=["no_enrollment",'Part time course','Full time course']
df['enrolled_university']=df['enrolled_university'].astype('category')
df['enrolled_university']=df['enrolled_university'].cat.set_categories(new_categories=univ_cat,ordered=True)
df['enrolled_university'].cat.categories

Index(['no_enrollment', 'Part time course', 'Full time course'], dtype='object')

In [51]:
df['relevant_experience']=df['relevant_experience'].apply(lambda x : bool(1) if x=='Has relevant experience' else bool(0) )

In [52]:
df['gender'].value_counts()

gender
Male      4994
Female     459
Other       75
Name: count, dtype: int64

In [53]:
gender_cat=['Male','Female','Other']
df['gender']=df['gender'].astype('category')
df['gender']=df['gender'].cat.set_categories(new_categories=gender_cat)
df['gender'].cat.categories

Index(['Male', 'Female', 'Other'], dtype='object')

In [54]:
df['student_id']=df['student_id'].astype('int32')
df['city_development_index']=df['city_development_index'].astype('float16')
df['training_hours']=df['training_hours'].astype('int32')
df['job_change']=df['job_change'].astype('bool')

In [55]:
df.head()

student_id      city  city_development_index gender  relevant_experience  \
0        8949  city_103                0.919922   Male                 True   
1       29725   city_40                0.775879   Male                False   
2       11561   city_21                0.624023    NaN                False   
3       33241  city_115                0.789062    NaN                False   
4         666  city_162                0.767090   Male                 True   

  enrolled_university education_level major_discipline experience  \
0       no_enrollment        Graduate             STEM        >20   
1       no_enrollment        Graduate             STEM         15   
2    Full time course        Graduate             STEM          5   
3                 NaN        Graduate  Business Degree         <1   
4       no_enrollment         Masters             STEM        >20   

  company_size    company_type last_new_job  training_hours  job_change  
0          NaN             NaN            1              36        True  
1        50-99         Pvt Ltd           >4              47       False  
2          NaN             NaN        never              83       False  
3          NaN         Pvt Ltd        never              52        True  
4        50-99  Funded Startup            4               8       False

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7142 entries, 0 to 7141
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   student_id              7142 non-null   int32   
 1   city                    7142 non-null   category
 2   city_development_index  7142 non-null   float16 
 3   gender                  5528 non-null   category
 4   relevant_experience     7142 non-null   bool    
 5   enrolled_university     6996 non-null   category
 6   education_level         6984 non-null   category
 7   major_discipline        6097 non-null   category
 8   experience              7125 non-null   category
 9   company_size            4918 non-null   category
 10  company_type            4851 non-null   category
 11  last_new_job            5904 non-null   category
 12  training_hours          7142 non-null   int32   
 13  job_change              7142 non-null   bool    
dtypes: bool(2), category(9),

In [57]:
df=df[(df['experience']>='10') & (df['company_size']>='1000-4999')]
df.shape
df.memory_usage().sum()

30499

In the beginning, the data memory usage was 800032, and now it is 30499. It dropped 26 times just by changing the column types.